core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/async/SplittingTransformer.java

```diff
@@ -0,0 +1,335 @@
+
+    /**
+     * The downstream subscriber that is subscribed to this publisher.
+     */
+    private Subscriber<? super AsyncResponseTransformer<ResponseT, ResponseT>> downstreamSubscriber;
+
+    /**
+     * The amount requested by the downstream subscriber that is still left to fulfill. Updated. when the
+     * {@link Subscription#request(long) request} method is called on the downstream subscriber's subscription. Corresponds to the
+     * number of {@code AsyncResponseTransformer} that will be published to the downstream subscriber.
+     */
+    private final AtomicLong outstandingDemand = new AtomicLong(0);
+
+    /**
+     * tracks if the publisher is currently sending AsyncResponseTransformer to the downstreamSubscriber
+     */
+    private final AtomicBoolean emitting = new AtomicBoolean(false);
+
+    // 'boundedness' info
+    private final long maxElements;
+    private final boolean isBounded;
+    private final AtomicInteger totalEmitted = new AtomicInteger(0);
+    private final AtomicBoolean terminated = new AtomicBoolean(false);
+
+    /**
+     * Creates a publisher bounded by the {@code maxElements} provided. As per reactive stream specification, this publisher will
+     * be considered 'unbounded' if {@code maxElements == LONG.MAX_VALUE}.
+     */
+    public SplittingTransformer(AsyncResponseTransformer<ResponseT, ResultT> upstreamResponseTransformer,
+                                long bufferSize, CompletableFuture<ResultT> returnFuture,
+                                long maxElements) {
+        this.upstreamResponseTransformer = Validate.paramNotNull(upstreamResponseTransformer, "asyncRequestBody");
+        this.returnFuture = Validate.paramNotNull(returnFuture, "returnFuture");
+        this.maximumBufferSize = NumericUtils.saturatedCast(bufferSize);
+        this.maxElements = maxElements;
+        this.isBounded = maxElements != Long.MAX_VALUE;
+    }
+
+    /**
+     * Creates an unbounded publisher.
+     */
+    public SplittingTransformer(AsyncResponseTransformer<ResponseT, ResultT> upstreamResponseTransformer,
+                                long bufferSize, CompletableFuture<ResultT> returnFuture) {
+        this.upstreamResponseTransformer = Validate.paramNotNull(upstreamResponseTransformer, "asyncRequestBody");
+        this.returnFuture = Validate.paramNotNull(returnFuture, "returnFuture");
+        this.maximumBufferSize = NumericUtils.saturatedCast(bufferSize);
+        // unbounded by default
+        this.maxElements = Long.MAX_VALUE;
+        this.isBounded = false;
+    }
+
+    /**
+     * @param downstreamSubscriber the {@link Subscriber} to the individual AsyncResponseTransformer
+     */
+    @Override
+    public void subscribe(Subscriber<? super AsyncResponseTransformer<ResponseT, ResponseT>> downstreamSubscriber) {
+        if (downstreamSubscriber == null) {
+            // §1.09 Subscribe throw NPE on null subscriber
```

Comment:
Is this comment needed? Same as other reactive streams related comments. Seems to clutter the code IMO